### PR TITLE
[SPARK-9952] Fix N^2 loop when DAGScheduler.getPreferredLocsInternal accesses cacheLocs

### PR DESCRIPTION
**Pull request by JoshRosen**
In Scala, `Seq.fill` always seems to return a List. Accessing a list by index is an O(N) operation. Thus, the following code will be really slow (~10 seconds on my machine):

``` scala
val numItems = 100000
val s = Seq.fill(numItems)(1)
for (i <- 0 until numItems) s(i)
```

It turns out that we had a loop like this in DAGScheduler code, although it's a little tricky to spot. In `getPreferredLocsInternal`, there's a call to `getCacheLocs(rdd)(partition)`.  The `getCacheLocs` call returns a Seq. If this Seq is a List and the RDD contains many partitions, then indexing into this list will cost O(partitions). Thus, when we loop over our tasks to compute their individual preferred locations we implicitly perform an N^2 loop, reducing scheduling throughput.

This patch fixes this by replacing `Seq` with `Array`.

### CODE REVIEW

**Comment by JoshRosen** [comment]
The problematic `Seq.fill` was added in #6352, which was merged to 1.5.0, so I think that this patch needs to be merged to master and 1.5.0 in order to prevent a scheduling performance regression.

**Comment by JoshRosen** [comment]
I noticed this while running a very simple scheduling throughput benchmark under YourKit Java profiler with CPU tracing enabled.  Here's a comparison of two trace results, clearly illustrating the slowdown:

image

For scheduling a job with 10000 no-op tasks, (`sc.makeRDD(1 to NUM_TASKS, NUM_TASKS).mapPartitions(identity).count()`), the end-to-end time in `local[*]` mode dropped from ~48 seconds to ~20 seconds as a result of this change.

**Comment by JoshRosen** [comment]
Also, note that the actual max scheduling throughput is much higher with tracing disabled; I can scheduler over 5000 tasks / second on my laptop.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by squito** [comment]
nice find.  We could also use an `IndexedSeq` (which I think is what everyone really wants anytime they say `Seq`), but it looks like there are other places that want an array anyway, so makes sense to just use an array here.  `IndexedSeq` would leave the door open for other optimizations -- eg., when the storage level is `None`, you don't actually need to create anything, you can just return a collection which knows its `Nil` in every slot ... but that would probably be premature optimization.

anyway, just some random thoughts.  lgtm pending tests.

**Comment by JoshRosen** [comment]
One potential gotcha of using arrays: we might run into problems with array equality checks returning `false` because they're based on the array identity rather than contents. Given this, maybe it would be safer to use `IndexedSeq`; the alternative would be to carefully search the existing code to figure out whether this change has accidentally broken any comparisons.

**Comment by JoshRosen** [comment]
I've gone ahead and minimized this to just the `IndexedSeq` change, which seems to address the performance issue, so I'm going to merge this to master and 1.5 pending Jenkins.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 1]
```diff
@@ -302,12 +302,12 @@ class DAGScheduler(
       shuffleDep: ShuffleDependency[_, _, _],
       firstJobId: Int): ShuffleMapStage = {
     val rdd = shuffleDep.rdd
-    val numTasks = rdd.partitions.size
+    val numTasks = rdd.partitions.length
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: super nit, but any reason for all the `size` -> `length` changes?  just seems like a bit of noise if we ever look in git history for these lines.

**Comment by squito** [comment]
thanks for updating josh.  still lgtm pending tests from me.  (left one minor comment, your discretion to update).

**Comment by squito** [comment]
Jenkins, retest this please

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala (hunk 1 above)
Comment (markhamstra) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: I'm not sure how true it is with more recent versions of Scala, but there at least was a time when `Array#size` didn't perform nearly as well as `Array#length`.

Comment (markhamstra) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: http://www.scala-lang.org/old/node/7650

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by andrewor14** [comment]
retest this please

**Comment by andrewor14** [comment]
LGTM will merge once we pass tests.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by rxin** [comment]
I'm going to merge this since the unit tests just took longer to run but actually the relevant tests passed.

**Pull request closed without merging**